### PR TITLE
SARSA Algo 

### PR DESCRIPTION
**Pull request by Het-Shah**
Implemented this with eligibility traces, hopefully this goes in classicalrl. For Issue #31

### CODE REVIEW

File: jigglypuffRL/classicalrl/sarsa/sarsa.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        self.max_episodes = max_episodes
+        self.epsilon = epsilon
+        self.decay_rate_epsilon = decay_rate_epsilon
+        self.alpha = alpha 
+        self.gamma = gamma
+        self.lmbda = lmbda
+        self.thres_reward = thres_reward
+        
+        # Set up the Q table 
+        self.Q_table = np.zeros((self.env.observation_space.n,self.env.action_space.n))
```
Comment (ajaysub110) [inline jigglypuffRL/classicalrl/sarsa/sarsa.py:41]: Should we initialize the Q_table with a random (kaiming uniform?) distribution instead of all zeros?

File: jigglypuffRL/classicalrl/sarsa/sarsa.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    def select_action(self, state):
+        # epsilon greedy method to sample actions
+        if np.random.uniform(0, 1) < self.epsilon:
+            action = self.env.action_space.sample()
+        else: 
+            action = np.argmax(self.Q_table[state, :])
+            
+        return action
+        
+    def update(self, r, state1, action1, state2, action2):
```
Comment (ajaysub110) [inline jigglypuffRL/classicalrl/sarsa/sarsa.py:54]: Can you call this update_params(),Let's keep the method names uniform for all algos

File: jigglypuffRL/classicalrl/sarsa/sarsa.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+                
+                state1 = state2 
+                action1 = action2
+
+                if done:
+                    self.epsilon = self.epsilon * self.decay_rate_epsilon
+                    break
+                
+            if ep % 5000 == 0:
+                #report every 5000 steps, test 100 games to get avarage point score for statistics and verify if it is solved
```
Comment (ajaysub110) [inline jigglypuffRL/classicalrl/sarsa/sarsa.py:85]: Won't this if condition be satisfied only on every 5000th episode? and not step?

File: jigglypuffRL/classicalrl/sarsa/sarsa.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                    while done != True: 
+                        action = np.argmax(self.Q_table[obs])
+                        obs, rew, done, info = self.env.step(action)
+                        rew_average += rew
+                rew_average=rew_average/100
+                print('Episode {} avarage reward: {}'.format(ep,rew_average))
+
+                if rew_average > self.thres_reward:
+                    print(f"{self.env.unwrapped.spec.id} solved")
+                    break
```
Comment (ajaysub110) [inline jigglypuffRL/classicalrl/sarsa/sarsa.py:100]: Can we remove this for now?

File: jigglypuffRL/classicalrl/sarsa/sarsa.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        self,
+        env,
+        max_iterations=1000,
+        max_episodes=1500001, 
+        epsilon=0.9,
+        alpha=0.1,
+        gamma=0.95,
+        lmbda=0.90,
+        decay_rate_epsilon=0.999,
+        thres_reward=0.8):
```
Comment (ajaysub110) [inline jigglypuffRL/classicalrl/sarsa/sarsa.py:28]: On what basis are we giving `thres_reward` here? Won't it vary between environments?

File: jigglypuffRL/classicalrl/sarsa/sarsa.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+                
+                for i in range(100):
+                    obs= self.env.reset()
+                    done=False
+                    while done != True: 
+                        action = np.argmax(self.Q_table[obs])
+                        obs, rew, done, info = self.env.step(action)
+                        rew_average += rew
+                rew_average=rew_average/100
+                print('Episode {} avarage reward: {}'.format(ep,rew_average))
```
Comment (ajaysub110) [inline jigglypuffRL/classicalrl/sarsa/sarsa.py:96]: typo in average. Actually you can replace `average reward` with just `Reward`

**Comment by ajaysub110** [comment]
Additionally, can you please add `tensorboard_log`, `render` and `seed` as params similar to what we have used in the `deeprl` algos?

**Comment by Het-Shah** [comment]
will add tensorboard logger

File: jigglypuffRL/classicalrl/sarsa/sarsa.py (hunk 4 above)
Comment (Het-Shah) [reply jigglypuffRL/classicalrl/sarsa/sarsa.py:100]: Removing thres_reward completely

File: jigglypuffRL/classicalrl/sarsa/sarsa.py (hunk 3 above)
Comment (Het-Shah) [reply jigglypuffRL/classicalrl/sarsa/sarsa.py:85]: yes this is what my idea was..

File: jigglypuffRL/classicalrl/sarsa/sarsa.py (hunk 2 above)
Comment (Het-Shah) [reply jigglypuffRL/classicalrl/sarsa/sarsa.py:54]: sorry about this one

File: jigglypuffRL/classicalrl/sarsa/sarsa.py (hunk 3 above)
Comment (Het-Shah) [reply jigglypuffRL/classicalrl/sarsa/sarsa.py:85]: to call it after 5000 episodes will change the comment

File: jigglypuffRL/classicalrl/sarsa/sarsa.py (hunk 1 above)
Comment (Het-Shah) [reply jigglypuffRL/classicalrl/sarsa/sarsa.py]: Regarding this, I tried this didn't seem to train properly.

Comment (ajaysub110) [reply jigglypuffRL/classicalrl/sarsa/sarsa.py]: Is it working well with all zeros? Would giving the user an option to choose initialization be a good idea? I don't know

File: jigglypuffRL/classicalrl/sarsa/sarsa.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    Paper: http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.17.2539&rep=rep1&type=pdf
+    :param env: (Gym environment) The environment to learn from
+    :param max_iterations: (int) Maximum number of iterations per episode
+    :param max_epsiodes: (int) Maximum number of episodes
+    :param epsilon: (float) Epsilon for epsilon-greedy selection of the action
+    :param alpha: (float) Learning rate
+    :param gamma: (float) Discount Factor
+    :param lmbda: (float) Lambda for the eligibility traces
+    :param decay_rate_epsilon: (float) Decay rate for the epsilon of epsilon greedy action selection
+    :param thres_reward: (float) Threshold of the reward you want the training to stop
```
Comment (ajaysub110) [inline jigglypuffRL/classicalrl/sarsa/sarsa.py:17]: Pls add the new params here to the docs as well.

File: jigglypuffRL/classicalrl/sarsa/sarsa.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        self.lmbda = lmbda
+        self.tensorboard_log = tensorboard_log
+        self.seed = seed
+        self.render = render
+        
+        # Assign seed
+        if seed is not None:
+            torch.manual_seed(seed)
+            torch.backends.cudnn.deterministic = True
+            torch.backends.cudnn.benchmark = False
```
Comment (ajaysub110) [inline jigglypuffRL/classicalrl/sarsa/sarsa.py:49]: Why is torch required?

File: jigglypuffRL/classicalrl/sarsa/sarsa.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+        self.writer = None
+        if self.tensorboard_log is not None:
+            from torch.utils.tensorboard import SummaryWriter
+
+            self.writer = SummaryWriter(log_dir=self.tensorboard_log)
+        
+        # Set up the Q table 
+        self.Q_table = np.zeros((self.env.observation_space.n,self.env.action_space.n)) 
+        # Set up eligibility traces
+        self.e_table = np.zeros((self.env.observation_space.n,self.env.action_space.n))
```
Comment (ajaysub110) [inline jigglypuffRL/classicalrl/sarsa/sarsa.py:62]: Please reformat your code with black before any merge

**Comment by Het-Shah** [comment]
made the changes

**Comment by ajaysub110** [comment]
Great, thanks! Merging
